doc/specs/#1595 - Suggestions UI/Suggestions-UI.md

```diff
@@ -0,0 +1,734 @@
+
+To drill in on a single example - the `commandHistory` source. In that
+particular case, the TerminalPage will query the active TermControl for a list
+of its recent commands. If it knows these (via shell integration), then the
+TerminalPage will use that list of commands to build a list of `sendInput`
+actions. Those will then get fed to the suggestions UI.
+
+Not listed above is [shell-driven autocompletion]. These aren't something that
+the Terminal can invoke all on its own - these are something the shell would
+need to invoke themselves.
+
+#### Pre-populate the current commandline context
+
+Consider the following scenario. A user has typed `git c` in their shell, and
+has [shell integration] enabled for their shell. They want to open the
+Suggestions UI filtered to their recent history, but starting with what they've
+already typed. To support this scenario, we'll add an additional property:
+
+* `"useCommandline"`: `bool` (**default**: `true`)
+  * `true`: the current commandline the user has typed will pre-populate the
+    filter of the Suggestions UI. This requires that the user has enabled shell
+    integration in their shell's config.
+  * `false`: the filter will start empty, regardless of what the user has typed.
+
+With that setting, the user can achieve their desired UX with the following action:
+
+```json
+{ "command": { "action":"suggestions", "source": "commandHistory", "useCommandline": true } },
+```
+
+Now, when they type `git c` and invoke the Suggestions UI, they can immediately
+start searching for recent commands that started with `git c`.
+
+The primary use case for `useCommandline: false` was for `"nesting": "source"`.
+When filtering a list of ["Tasks...", "Recent commands...", "Recent
+directories...", "Docker...", "Git..."], then there's minimal value to start by
+filtering to "git c".
+
+#### Default actions
+
+I propose adding the following actions to the Terminal by default:
+
+```json
+{ "command": { "action":"suggestions", "source": "commandHistory", "useCommandline": true } },
+{ "command": { "action":"suggestions", "source": "directoryHistory" } },
+{ "command": { "action":"suggestions", "source": ["local", "tasks", "commandHistory"], "useCommandline": true, "nesting": "disabled" } },
+{ "command": { "action":"suggestions", "source": ["all"], "useCommandline": false, "nesting": "source" } },
+```
+
+These actions are colloquially:
+* Give me suggestions from my recent commands, using what I've typed
+* Give me suggestions of directories I've recently been in
+* _(After [Tasks] are implemented)_ Give me suggestions from recent commands,
+  commands I've saved, and commands for this project. Don't nest any, so they're
+  all in the top-level menu. Use what I've typed already to start filtering. 
+* Just open the Suggestions UI with all suggestions sources, and group them by
+  the source of the suggestions. 
+
+This should cover most of the basic use cases for suggestions. 

```

Comment:
Any rulings on default bindings?

Reply:
This isn't something we've really discussed (or really done a whole lot of "default" bindings for recently). 

Maybe Ctrl+Shift+S for the

```json
{ "command": { "action":"suggestions", "source": ["local", "tasks", "commandHistory"], "useCommandline": true, "nesting": "disabled" } },
```

one? The "give me all the possible commandlines" one?

I've been personally using `Ctrl+Shift+H` and `Ctrl+Shift+Y` because they're convenient with the moonlander.